README.md
<!-- BACK TO TOP LINK -->

<a name="readme-top"></a>

<!-- PROJECT SHIELDS -->

[![Forks][Forks Shield]][Forks URL]
[![Stargazer][Stars Shield]][Stars URL]
[![MIT License][License Shield]][License URL]

# semantic-release

semantic-release는 버전 관리를 자동화할 수 있도록 도움을 주는 라이브러리입니다. 이를 사용하면 커밋 타입이나 메시지 내용에 따라 프로젝트 버전을 증가시킬 수도 있고, Release 노트 작성을 자동화할 수도 있습니다. 저는 회사 프로젝트에 적용해 봤는데 🐶 만족 중입니다.

### 워크플로우

- 하위 브랜치에서 작업 후 변경사항을 커밋합니다.
  - semantic-release의 커밋 규칙은 [conventional commits](https://github.com/conventional-changelog/commitlint/tree/master/%40commitlint/config-conventional)를 따릅니다.
- 기능 구현이 완료되었다면, main 브랜치에 병합합니다.
- 커밋 타입 또는 메시지 내용에 따라 버전이 관리되고, 릴리즈 노트가 자동으로 작성됩니다.
  - 커밋 메시지 내용에 `breaking change`가 포함되어 있다면, `major` 버전이 증가합니다.
  - 커밋 타입이 `feat`로 시작한다면, `minor` 버전이 증가합니다.
  - 커밋 타입이 `fix` 또는 `perf`로 시작한다면, `patch` 버전이 증가합니다.

> 버전은 시맨틱 버저닝(major.minor.patch)을 따르고 있으며, 앞의 규칙부터 버전 관리 시 우선 적용됩니다.
> 병합 하기 전의 커밋 내역에 feat와 fix가 모두 있다면, minor 버전만 증가하는 식입니다.

### 설치

```bash
$ yarn add -D semantic-release @semantic-release/changelog @semantic-release/git
```

### 설정

루트 경로에 `release.config.js` 또는 `.releaserc` 파일을 생성하세요.

```javascript
/**
 * Semantic Release
 * - Commit 타입에 따라 릴리즈 노트를 작성해 주는 패키지입니다.
 * - https://github.com/semantic-release/semantic-release
 */
const config = {
  branches: ['main'],
  plugins: [
    [
      '@semantic-release/github',
      {
        successComment: false,
        failTitle: false,
      },
    ],
    '@semantic-release/commit-analyzer',
    '@semantic-release/release-notes-generator',
    '@semantic-release/changelog',
    [
      '@semantic-release/npm',
      {
        npmPublish: false,
      },
    ],
    [
      '@semantic-release/git',
      {
        message: 'chore(release): v${nextRelease.version}\n\n${nextRelease.notes}',
      },
    ],
  ],
};

module.exports = config;
```

앞서 설치하지 않은 플러그인이 일부 추가된 것을 볼 수 있는데, 이들은 semantic-release에 기본적으로 내장된 플러그인입니다. 각각 역할이 다르니 이는 공식문서에서 확인해 주세요. 그리고 semantic-release는 동작 순서가 있으며, 각 플러그인이 의존하는 순서가 있기 때문에 포함되는 순서를 신경 써야 합니다.

### GitHub Action CI/CD 구성

`/.github/workflows` 경로 안에 `.yml` 파일을 생성하세요.

```yml
name: Release New Version
on:
  push:
    branches:
      - main
jobs:
  release:
    name: Release
    runs-on: ubuntu-latest
    permissions:
      contents: write
      issues: write
      pull-requests: write
    steps:
      - name: Checkout
        uses: actions/checkout@v3
      - name: Setup Node.js
        uses: actions/setup-node@v3
        with:
          node-version: 'lts/*'
      - name: Install dependencies
        run: rm -rf node_modules && yarn install --frozen-lockfile
      # - name: Build application
      #   run: yarn build
      - name: Run semantic-release
        run: npx semantic-release
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

main 브랜치에 병합될 때만 Action이 동작하도록 작성했습니다. 리액트와 같은 프로젝트를 진행하고 계시다면, 주석 처리된 부분을 해제하여, 빌드도 진행하시면 좋습니다. 만약 빌드가 제대로 되지 않는 경우, release 또한 완료되지 않기에 버전 관리를 안전하게 진행할 수 있습니다.

<br />

# cz-customizable

cz-customizable은 커밋 규칙을 강제하기 위해 도입한 패키지입니다. 팀 내에서 규칙을 정했다 해도, 항상 매뉴얼로 작성하다 보면 헷갈릴 때가 있습니다. 이 경우 사용하면 좋을 거 같아 적용했는데, semantic-release와 함께 사용할 때 경험이 꽤 좋았습니다.

### 설치

```bash
$ yarn add -D cz-customizable
```

### 설정

루트 경로에 `cz-config.js` 파일을 생성하세요.

```javascript
module.exports = {
  types: [
    { value: 'feat', name: 'feat: 새로운 기능 추가' },
    { value: 'fix', name: 'fix: 잘못된 동작이나 버그 수정' },
    { value: 'perf', name: 'perf: 성능 개선' },
    { value: 'refactor', name: 'refactor: 코드 리팩토링' },
    { value: 'revert', name: 'revert: 이전의 코드로 되돌림' },
    { value: 'style', name: 'style: 포맷팅이나 컨벤션에 따른 수정 사항 반영' },
    { value: 'test', name: 'test: 테스트 코드 추가 & 리팩토링' },
    { value: 'build', name: 'build: 빌드 관련 코드 추가 & 수정' },
    { value: 'ci', name: 'ci: CI 업데이트' },
    { value: 'docs', name: 'docs: 문서 작성 & 수정' },
    { value: 'chore', name: 'chore: 사소한 변경이나 패키지매니저 관리' },
  ],
  scopes: [
    'component',
    'css-style',
    'custom-hook',
    'store',
    'util',
    'api',
    'assets',
    'package',
    'lint',
    'formatting',
    'config',
    'workflow',
    'README',
  ],
  allowCustomScopes: true,
  allowBreakingChanges: ['feat', 'fix'],
  messages: {
    type: '✨ Commit 타입 선택:',
    scope: '✨ Commit 범위 선택 (선택사항):',
    customScope: '✨ Custom 범위 작성:\n',
    subject: '✨ Commit 제목 작성:\n',
    body: '✨ Commit 내용 작성 (선택사항) ["|" 사용 시 개행 처리]:\n',
    breaking: '✨ BREAKING CHANGES 사항 나열 (선택사항):\n',
    footer: '✨ 이 변경으로 끝난 이슈 나열 (선택사항). E.g.: #31, #34:\n',
    confirmCommit: '✨ 작성한 내용으로 커밋을 진행하겠습니까?',
  },
  skipQuestions: ['body'],
  subjectLimit: 100,
};
```

`package.json`에 스크립트와 설정을 추가하세요.

```json
{
  ...
  "scripts": {
    "commit": "./node_modules/cz-customizable/standalone.js"
  },
  "config": {
    "cz-customizable": {
      "config": "cz-config.js"
    }
  },
  ...
}
```

변경사항을 Stage에 추가한 후, `yarn commit` 명령으로 실행하면 됩니다.

![Jun-26-2023 21-44-20](https://github.com/n0hack/semantic-release/assets/42988225/6676aac1-9eb7-4645-80fb-2ca0dd508381)

<p align="right">
  <a href="#readme-top">back to top</a>
</p>

<!-- MARKDOWN LINKS & IMAGES -->

[Forks Shield]: https://img.shields.io/github/forks/n0hack/semantic-release?style=for-the-badge
[Forks URL]: https://github.com/n0hack/semantic-release/network/members
[Stars Shield]: https://img.shields.io/github/stars/n0hack/semantic-release?style=for-the-badge
[Stars URL]: https://github.com/n0hack/semantic-release/stargazers
[License Shield]: https://img.shields.io/github/license/n0hack/semantic-release?style=for-the-badge
[License URL]: https://github.com/n0hack/semantic-release/blob/main/LICENSE


CHANGELOG.md
## [4.0.1](https://github.com/n0hack/semantic-release/compare/v4.0.0...v4.0.1) (2023-06-26)


### Bug Fixes

* 코드 수정 ([44acdcc](https://github.com/n0hack/semantic-release/commit/44acdcc38b688620c49441e83f695c9e7a517f6b))

# [4.0.0](https://github.com/n0hack/semantic-release/compare/v3.4.0...v4.0.0) (2023-06-26)


### Bug Fixes

* **config:** f ([cfd4f84](https://github.com/n0hack/semantic-release/commit/cfd4f843bd643d696bce645ad8744e5f24a631ff))


### Features

* **component:** 테스트 ([21b7b91](https://github.com/n0hack/semantic-release/commit/21b7b91030861637bd183485048eb669836ab725))
* **커스텀:** 커스텀 커밋 ([5172ed3](https://github.com/n0hack/semantic-release/commit/5172ed31434e3c6e87f5311e6b1c5d441ddbcabc)), closes [#55](https://github.com/n0hack/semantic-release/issues/55)


### BREAKING CHANGES

* **config:** z

ISSUES CLOSED: 77
* **커스텀:** ok

# [3.4.0](https://github.com/n0hack/semantic-release/compare/v3.3.0...v3.4.0) (2023-06-26)


### Features

* **no-release:** 추가 ([2837d67](https://github.com/n0hack/semantic-release/commit/2837d67a694b65c4442a8fd1163ca3990fda5320))

# [3.3.0](https://github.com/n0hack/semantic-release/compare/v3.2.0...v3.3.0) (2023-06-26)


### Bug Fixes

* **LiL:** laLa ([0116e2c](https://github.com/n0hack/semantic-release/commit/0116e2c6cac8ae231bb2a4af7eeb7e1bf4f65344))


### Features

* **Lint:** Commitzen ([e2a5be0](https://github.com/n0hack/semantic-release/commit/e2a5be0643bc4c70fb60b9b8163072bdde0b4cf5))

# [3.2.0](https://github.com/n0hack/semantic-release/compare/v3.1.2...v3.2.0) (2023-06-26)


### Features

* **scope:** 이건 ? ([b84508a](https://github.com/n0hack/semantic-release/commit/b84508ad2f5d48f2daa88a97d6f9caf31888f0dd))

## [3.1.2](https://github.com/n0hack/semantic-release/compare/v3.1.1...v3.1.2) (2023-06-26)


### Performance Improvements

* perf ([05e86f0](https://github.com/n0hack/semantic-release/commit/05e86f006b4d9111e36e685da77838442921e7b8))

## [3.1.1](https://github.com/n0hack/semantic-release/compare/v3.1.0...v3.1.1) (2023-06-26)


### Bug Fixes

* zz ([0e9b4e9](https://github.com/n0hack/semantic-release/commit/0e9b4e9fcd674dd8339cbab63ace669dba1b7e21))

# [3.1.0](https://github.com/n0hack/semantic-release/compare/v3.0.0...v3.1.0) (2023-06-26)





# [v3.1.0](https://github.com/n0hack/semantic-release/compare/v3.0.0...v3.1.0) (2023-06-26)

## ✨ New Features
- [`0a59aa3`](https://github.com/n0hack/semantic-release/commit/0a59aa3)  feat: 기능 추가 1 
- [`13b2e01`](https://github.com/n0hack/semantic-release/commit/13b2e01)  feat: 기능 추가 2

# [3.0.0](https://github.com/n0hack/semantic-release/compare/v2.0.0...v3.0.0) (2023-06-26)





# [v3.0.0](https://github.com/n0hack/semantic-release/compare/v2.0.0...v3.0.0) (2023-06-26)

## 💥 Breaking Changes
- [`1d54338`](https://github.com/n0hack/semantic-release/commit/1d54338)  feat: zzz

# [2.0.0](https://github.com/n0hack/semantic-release/compare/v1.4.0...v2.0.0) (2023-06-26)


### Bug Fixes

* 버그 수정 ([4baeb73](https://github.com/n0hack/semantic-release/commit/4baeb73e4e7daeba14b80cfa32ceaa46dcd0b534))


### BREAKING CHANGES

* 버전 올리장

# [1.4.0](https://github.com/n0hack/semantic-release/compare/v1.3.0...v1.4.0) (2023-06-26)


### Features

* 브챈 BREAKING CHANGE ([f0e8ec3](https://github.com/n0hack/semantic-release/commit/f0e8ec30bb122244ea08438f8ef75b72813fcb1e))

# [1.3.0](https://github.com/n0hack/semantic-release/compare/v1.2.0...v1.3.0) (2023-06-26)


### Bug Fixes

* 수정 ([a9f09c2](https://github.com/n0hack/semantic-release/commit/a9f09c21d136ddde7f77f92b68c604a98360c22e))


### Features

* 기능 추가 ([30e0ab6](https://github.com/n0hack/semantic-release/commit/30e0ab6d53d22c12643e3297a6fc7813081c256f))

# [1.2.0](https://github.com/n0hack/semantic-release/compare/v1.1.0...v1.2.0) (2023-06-26)


### Features

* 333333333 ([60db733](https://github.com/n0hack/semantic-release/commit/60db7332946892468d3b7bcfea903f03fce85c6e))
* ddddddd ([7400595](https://github.com/n0hack/semantic-release/commit/7400595b354f2ef431442bee040d29cf15197147))

# [1.1.0](https://github.com/n0hack/semantic-release/compare/v1.0.0...v1.1.0) (2023-06-26)


### Features

* f2 ([f339ae3](https://github.com/n0hack/semantic-release/commit/f339ae33d5f7f8373b820ecd24c8f53bd4620c89))

# 1.0.0 (2023-06-26)


### Features

* app ([277e3f3](https://github.com/n0hack/semantic-release/commit/277e3f313c5f4f8e4f6d65dbec7c0b4e43d31fec))


cz-legacy.js
module.exports = {
  types: [
    { value: 'feat', name: 'Feat: 기능을 추가해요.' },
    { value: 'fix', name: 'Fix: 버그를 수정해요.' },
    { value: 'perf', name: 'Perf: 성능을 개선해요.' },
    { value: 'refactor', name: 'Refactor: 코드를 리팩토링해요.' },
    { value: 'style', name: 'Style: 포맷팅이나 컨벤션에 따라 수정 사항을 반영해요.' },
    { value: 'docs', name: 'Docs: 문서의 내용을 일부 변경해요.' },
    { value: 'test', name: 'Test: 테스트 코드를 추가하거나 리팩토링해요.' },
    { value: 'chore', name: 'Chore: 사소한 변경사항이나, 패키지매니저를 관리해요.' },
    { value: 'revert', name: 'Revert: 이전의 코드로 되돌려요.' },
    { value: 'move', name: 'Move: 디렉토리, 파일이나 코드를 새로운 위치로 이동시켜요.' },
    { value: 'remove', name: 'Remove: 쓸모없는 디렉토리, 파일이나 코드를 삭제해요.' },
    { value: 'ci', name: 'CI: CI를 업데이트해요.' },
  ],
  scopes: [
    'component',
    'css-style',
    'custom-hook',
    'store',
    'util',
    'api',
    'wrong codes',
    'spaghetti codes',
    'alien codes',
    'assets',
    'package',
    'lint',
    'formatting',
    'config',
    'workflow',
    'breaking',
    'no-release',
    'README',
  ],
  allowCustomScopes: true,
};

// build: Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
// ci: Changes to our CI configuration files and scripts (examples: CircleCi, SauceLabs)
// docs: Documentation only changes
// feat: A new feature
// fix: A bug fix
// perf: A code change that improves performance
// refactor: A code change that neither fixes a bug nor adds a feature
// test: Adding missing tests or correcting existing tests


cz-config.js
module.exports = {
  types: [
    { value: 'feat', name: 'feat: 새로운 기능 추가' },
    { value: 'fix', name: 'fix: 잘못된 동작이나 버그 수정' },
    { value: 'perf', name: 'perf: 성능 개선' },
    { value: 'refactor', name: 'refactor: 코드 리팩토링' },
    { value: 'revert', name: 'revert: 이전의 코드로 되돌림' },
    { value: 'style', name: 'style: 포맷팅이나 컨벤션에 따른 수정 사항 반영' },
    { value: 'test', name: 'test: 테스트 코드 추가 & 리팩토링' },
    { value: 'build', name: 'build: 빌드 관련 코드 추가 & 수정' },
    { value: 'ci', name: 'ci: CI 업데이트' },
    { value: 'docs', name: 'docs: 문서 작성 & 수정' },
    { value: 'chore', name: 'chore: 사소한 변경이나 패키지매니저 관리' },
  ],
  scopes: [
    'component',
    'css-style',
    'custom-hook',
    'store',
    'util',
    'api',
    'assets',
    'package',
    'lint',
    'formatting',
    'config',
    'workflow',
    'README',
  ],
  allowCustomScopes: true,
  allowBreakingChanges: ['feat', 'fix'],
  messages: {
    type: '✨ Commit 타입 선택:',
    scope: '✨ Commit 범위 선택 (선택사항):',
    customScope: '✨ Custom 범위 작성:\n',
    subject: '✨ Commit 제목 작성:\n',
    body: '✨ Commit 내용 작성 (선택사항) ["|" 사용 시 개행 처리]:\n',
    breaking: '✨ BREAKING CHANGES 사항 나열 (선택사항):\n',
    footer: '✨ 이 변경으로 끝난 이슈 나열 (선택사항). E.g.: #31, #34:\n',
    confirmCommit: '✨ 작성한 내용으로 커밋을 진행하겠습니까?',
  },
  skipQuestions: ['body'],
  subjectLimit: 100,
};
